IBM MQ Filebeat module

felix-lessoer opened the pull request:
This module is able to read errorlogs from IBM MQ.

@ruflin supporting this.

ruflin commented:
Still need to have a closer look. Two things so far:

* Could you add an entry to the changelog?
* Could you add some example logs to the tests directory?

felix-lessoer commented on filebeat/docs/modules/ibmmq.asciidoc:
```diff
@@ -0,0 +1,56 @@
+
+This module comes with a sample dashboard. For example:
+
+TODO: include an image of a sample dashboard
```
![image](https://user-images.githubusercontent.com/35069463/58169384-e58f0100-7c90-11e9-9fac-3c36c63ceaff.png)

felix-lessoer replied:
I added this dashboard as kibana export

ruflin replied:
I added this to the PR and linked it to the docs.

felix-lessoer commented on x-pack/filebeat/module/ibmmq/errorlog/_meta/fields.yml:
```diff
@@ -0,0 +1,80 @@
+  - name: vrmf
+    description: Please add description
+    example: Please add example
+    type: text
```
VRMF (Version, Release, Maintenance, and Fix) is the version of the MQ instance

felix-lessoer replied:
Exactly

ruflin replied:
changed.

felix-lessoer commented on x-pack/filebeat/module/ibmmq/errorlog/_meta/fields.yml:
```diff
@@ -0,0 +1,80 @@
+    type: text
+  - name: iso8601_timestamp
+    description: Please add description
+    example: Please add example
```
Lets define it as date. Its the time when the log message was written into the logs. So when the events actually occurs.

felix-lessoer commented on x-pack/filebeat/module/ibmmq/errorlog/_meta/fields.yml:
```diff
@@ -0,0 +1,80 @@
+    description: Please add description
+    example: Please add example
+    type: text
+  - name: arithinsert1
```
The meaning is different based on the error code. Is there any way to reflect this? Same for comment fields.

felix-lessoer replied:
Well its possibly one per error code. And I didn't found a documentation for that on IBM side. 
So weit can ignore that fact like we do in Winlogbeat.
We can talk to IBM directly and see weather they can share more insights.
Or we can try to build our own documentation. Which would be hard I guess.
I will see what I can get from IBM and otherwise I think we need to stay with the first option. This fields mostly do not add relevant info as the values are part of the message anyways.

ruflin replied:
What about merging all artinsert into 1 field and the commentinsert into 1 field. Each field in Elasticsearch is an array, so the values would look like `arthinsert: ["val1", "val2"]`. Same for commentinsert. The main difference is that afterwards no query can be made "only" on arthinsert2. Would that be an issue?

felix-lessoer replied:
Well so far I have seen only one case where these information might be interesting. So I think its okay to do it like you suggested it.

ruflin replied:
I put all together for now.

felix-lessoer commented on x-pack/filebeat/module/ibmmq/errorlog/_meta/fields.yml:
```diff
@@ -0,0 +1,80 @@
+  - name: process
+    description: Please add description
+    example: Please add example
+    type: text
```
I'm not 100% sure about that, but it looks like the process ID (PID) in the host system of the IBM MQ sub program that is responsible for the error.

felix-lessoer replied:
It is PID and thread ID

ruflin replied:
moved it to ECS so the description here becomes obsolete.

felix-lessoer commented on x-pack/filebeat/module/ibmmq/errorlog/_meta/fields.yml:
```diff
@@ -0,0 +1,80 @@
+- name: errorlog
```
I created this sheet to make the mapping
https://docs.google.com/spreadsheets/d/16N85NC1pBctZ_qdwAqqIp520uw1hzW4-2gqG7YTL1yU/edit?usp=sharing

webmat replied:
Thanks for the great work, @felix-lessoer 👍 

Entered a few suggestions in the sheet.

felix-lessoer replied:
@ruflin I'm okay with that

ruflin replied:
adjusted to the sheet (were applicable)

webmat commented on x-pack/filebeat/module/ibmmq/errorlog/test/AMQERR01.log:
```diff
@@ -0,0 +1,356 @@
+
+EXPLANATION:
+An error has been detected, and the IBM MQ error recording routine has been
+called. The failing process is process 7360.
```
@ruflin Yeah I think that "process" may be pid + thread_id. Here, the error message refers to "process 7360", and the structured message above contains "7360.1".

webmat commented on x-pack/filebeat/module/ibmmq/errorlog/test/AMQERR01_QM2.log-expected.json:
```diff
@@ -0,0 +1,2758 @@
+        "log.flags": [
+            "multiline"
+        ],
+        "log.level": "",
```
The text is the same, though. I wonder if it's always the same or if in some cases it differs

webmat commented:
@ruflin Thanks for pinging me on this!

Don't wait for me for a review of the PR. But I'll definitely be coming back to this module, to see patterns for doing structured logging

ruflin commented on x-pack/filebeat/module/ibmmq/errorlog/test/AMQERR01.log-expected.json:
```diff
@@ -0,0 +1,377 @@
+    {
+        "@timestamp": "2018-10-11T08:46:25.924Z",
+        "ecs.version": "1.0.0",
+        "error.id": "AMQ7125I",
```
@webmat This brings up our old discussion around what should go into error and what not? Ingestion error or things from the error log.

Interestingly adding this field broke our tests as we check, that the error object does not exist.

webmat replied:
Well so far, all concrete uses of `error` have been for pipeline errors.

Accordingly, most error sources have been using `message` for their message. Although looking at all of the different text fields with different shades of error messages (description, explanation, etc) just below here, I'm not sure we can pick one directly for `message`. Is there one of them that's always there?

Anyhow, on to your actual question. In keeping with the actual convention, I'd say we shouldn't use `error.id` here, at least for now. The "userland" or source message going to message, I think the error code should go to `event.code`.

I'm saying this more to follow the current usage convention, more than declaring how it should be used in ECS. I think we can open up that discussion later, and see if we flesh out the usage of `error.*` and put pipeline meta-information such as errors elsewhere in one go, later. One other example of things we could support better is exceptions, stack traces & so on. I think all of that could eventually go in `error.*`, once it's fleshed out.

felix-lessoer replied:
All of them are always in the original message. 
The Description is the most general description of the error. The explanation is more deep. The action describes what a user can do to solve the error.

ruflin replied:
I put it under `ibmmq.errorlog.code` for now.

ruflin commented on filebeat/tests/system/test_modules.py:
```diff
@@ -162,8 +162,8 @@ def run_on_file(self, module, fileset, test_file, cfgfile):
 
-            assert "error" not in obj, "not error expected but got: {}".format(
-                obj)
+            #assert "error" not in obj, "not error expected but got: {}".format(
```
This is currently commented out to allow for `error.id`. But not sure we should store it into this value, it depends a bit on what we want to use it for long term in the context of ECS. @webmat

ruflin commented on x-pack/filebeat/module/ibmmq/_meta/docs.asciidoc:
```diff
@@ -0,0 +1,43 @@
+[float]
+=== Compatibility
+
+TODO: document with what versions of the software is this tested
```
@felix-lessoer Can you provide these details here?

felix-lessoer replied:
I tested in v 9.1.0.0 .. But the logs are the same in older versions

exekias replied:
I think we can just write that, it's good to explain what versions have we tested against, even if it's expected to work with others

ruflin commented on x-pack/filebeat/module/ibmmq/_meta/kibana/7/dashboard/Filebeat-IBMMQ-Overview.json:
```diff
@@ -0,0 +1,814 @@
+{
```
This dashboard will probably need an update as soon as we finished the field renaming.

ruflin commented on x-pack/filebeat/module/ibmmq/errorlog/_meta/fields.yml:
```diff
@@ -0,0 +1,42 @@
+  type: group
+  fields:
+  - name: installation
+    description: Please add description
```
@felix-lessoer I need some details here on the description.

felix-lessoer replied:
This is the installation name which can be given  at installation time. Default: Installation 1

From IBM documentation

Each installation of IBM® MQ on UNIX, Linux, and Windows, has a unique identifier known as an installation name. The installation name is used to associate things such as queue managers and configuration files with an installation.

ruflin replied:
updated. I wonder if we should in some cases even provide links to the docs.

ruflin commented on x-pack/filebeat/module/ibmmq/errorlog/_meta/fields.yml:
```diff
@@ -0,0 +1,42 @@
+    description: Please add description
+    example: Please add example
+    type: keyword
+  - name: qmgr
```
@felix-lessoer Can I get some details on what qmgr is?

felix-lessoer replied:
This is the name of the queue manager where the error has occurred on. 

From IBM documentation
Queue managers provide queuing services to applications, and manages the queues that belong to them. A queue manager ensures the following actions:
Object attributes are changed according to the commands received.
Special events such as trigger events or instrumentation events are generated when the appropriate conditions are met.
Messages are put on the correct queue, as requested by the application making the MQPUT call. The application is informed if this cannot be done, and an appropriate reason code is given.

ruflin replied:
updated.

ruflin commented on x-pack/filebeat/module/ibmmq/errorlog/ingest/pipeline.json:
```diff
@@ -0,0 +1,69 @@
+            "remove": {
+                "field": [
+                    "log_timestamp",
+                    "message"
```
Note for myself: Should we really fully remove message or keep at least something inside?

ruflin commented:
@felix-lessoer I have a more high level question: We call this fileset `errorlog`. Are there also other logs from ibmmq that we should add in the future? Or is this the only log and should just be `ibmmq.log`?

felix-lessoer commented:
Thats fine. Lets call its log for now

felix-lessoer commented on x-pack/filebeat/module/ibmmq/errorlog/_meta/fields.yml:
```diff
@@ -0,0 +1,34 @@
+    description: >
+      Name of the queue manager. Queue managers provide queuing services to applications, and manages the queues that belong to them.
+    type: keyword
+  - name: arithinsert
```
Well arithinsert1 is the first numeric value found in the error. So its hard to provide a better description.
Any ideas from your side?

ruflin replied:
How are arithinsert and commentinsert values different? How are they used?

`errordescription` is just a description of the log line / error?

felix-lessoer commented:
I updated the pipeline to handle error when fields are missing. I also added the @timestamp and message field to be able to use the default configuration of the Logs UI .
I also updated to sample Dashboard to fit to ECS . The Dashboard Import JSON looks the same as the others. But when I run `filebeat setup` I run into an error. No sure why.

exekias commented:
Checked the dashboard and it's now loading properly, we need to add filters to make sure that showed metrics and saved search only apply to data coming out of this module (`event.dataset: ibmmq.errorlog`).

exekias commented:
I've updated expected files, let's wait for CI :crossed_fingers:

exekias commented:
Tests are failing, it seems `arithinsert1` doesn't get removed for the sample log `AMQERR01.log`. @felix-lessoer I guess the pipeline is failing for that one?

exekias commented on x-pack/filebeat/module/ibmmq/errorlog/test/AMQERR01.log-expected.json:
```diff
@@ -0,0 +1,377 @@
+        "fileset.name": "errorlog",
+        "host.hostname": "FELIX-ELASTIC",
+        "ibmmq.errorlog.action": "None.",
+        "ibmmq.errorlog.arithinsert1": "90",
```
this field should be removed, :thinking:

exekias replied:
there was an error in the grok pattern, I fixed it

exekias commented:
failing tests are unrelated

ruflin commented:
🎉 Thanks @felix-lessoer and @exekias for pushing this through!
